pkg/manifests/manifests_test.go

```diff
@@ -3365,12 +3364,33 @@ nodeExporter:
 
 }
 
+func compareResources(gotResources, configResources *v1.ResourceRequirements) error {
+	if _, ok := configResources.Limits[v1.ResourceCPU]; ok {
+		if gotResources.Limits[v1.ResourceCPU] != configResources.Limits[v1.ResourceCPU] {
+			return fmt.Errorf("CPU limit incorrectly configured")
+		}
+	}
+	if _, ok := configResources.Limits[v1.ResourceMemory]; ok {
+		if gotResources.Limits[v1.ResourceMemory] != configResources.Limits[v1.ResourceMemory] {
+			return fmt.Errorf("memory limit incorrectly configured")
+		}
+	}
+	if _, ok := configResources.Requests[v1.ResourceCPU]; ok {
+		if gotResources.Requests[v1.ResourceCPU] != configResources.Requests[v1.ResourceCPU] {
+			return fmt.Errorf("CPU request incorrectly configured")
+		}
+	}
+	if _, ok := configResources.Requests[v1.ResourceMemory]; ok {
+		if gotResources.Requests[v1.ResourceMemory] != configResources.Requests[v1.ResourceMemory] {
+			return fmt.Errorf("memory request incorrectly configured")
+		}
+	}
+	return nil
+}
+
 func TestKubeStateMetrics(t *testing.T) {
 	config := `kubeStateMetrics:
   resources:
-    requests:
-      cpu: 100m
-      memory: 100Mi
```

Comment:
`requests` here will be now replaced with defaults, in case only `limits` were specified.

@machine424 WDYT?

Reply:
Yep, merging the config may be more practical for users.

We'll just need to see at which level should the merging stop (if the user only provides `requests.cpu` e.g. should we keep `defaultRequests.memory` or set it no nil?)

Maybe we can check for the behavior that the other operators opted for, to keep consistency.

Maybe we can ask the others for their opinions.

Reply:
It seems PO adheres to [the same approach](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/k8sutil/merge.go#L56), i.e., only overriding `requests.cpu` (if the user specified just that) while using the default value for `requests.memory`.

Reply:
We currently overwrite any existing defaults, even if the user only sets limits. So in that sense the merge proposal is functionality change. We should think through all upgrade scenarios and make sure nothing obviously bad can happen.
For example if a user has set a limit lower then the default request, this might work for them currently. If we now start merging the `Resources` fields, they will run into very annoying errors when they upgrade.

I think my preference would be to keep the old behavior (overwrite everything) and rather document this explicitly.

Reply:
I didn't know (or maybe I forgot) we already provide to customize resources for some components.
I don't know how much that customization is used, but apparently the two scenarios I mentioned don't seem to happen a lot (or we don't know about it).
Yep of course it's better to keep the same logic.

Reply:
Removed the latest commit, the PR implements the overriding behaviour now.